### PR TITLE
Implemented cleanup

### PR DESCRIPTION
**Pull request by CSVdB**

### CODE REVIEW

File: wolf-data/src/Wolf/Data.hs [hunk 1]
```diff
@@ -6,6 +6,10 @@ module Wolf.Data
     , initWolf
     , withInitCheck
     , withInitCheck_
+    -- * Prompt
+    , promptYesNo
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data.hs]: Wolf doesn't need to export these prompts. This is not wolf data functionality.
There is a reason this is a seperate import, and this is definitely a seperate refactor and not to be done here.

File: wolf-data/src/Wolf/Data.hs [hunk 2]
```diff
@@ -105,8 +110,23 @@ module Wolf.Data
     , exportRepo
     , importRepo
     , cleanupRepo
+    , Cautious(..)
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data.hs]: Wolf does not need to export Cautious.

File: wolf-data/src/Wolf/Data/Cautious.hs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+checkNotePersonRelation ::
+       (MonadIO m, MonadReader DataSettings m)
+    => PersonUuid
+    -> NoteUuid
+    -> CautiousS m ()
+checkNotePersonRelation pu nu = do
+    maybeNote <- lift $ readNote nu
+    case maybeNote of
+        Nothing -> cautiousWarning [WarnMissingNote nu] ()
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Cautious.hs]: Warning with a single warning may warrant a helper function.

File: wolf-data/src/Wolf/Data/Export.hs [hunk 4]
```diff
@@ -5,51 +5,67 @@ module Wolf.Data.Export
...
+getMapCautious ::
+       (Ord a, Monad m)
+    => (a -> CautiousS m (Maybe b))
+    -> [a]
+    -> CautiousS m (Map a b)
+getMapCautious func ls =
+    M.fromList . catMaybes . fmap sequence <$>
+    traverse (\a -> (,) a <$> func a) ls
+
+mKeyed :: (Ord a, Monad m) => (a -> m (Maybe b)) -> [a] -> m (Map a b)
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export.hs]: If you're only using this in `exportRepo`, put it in a `where` clause there.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 5]
```diff
@@ -4,13 +4,29 @@
...
+    , Cautious(..)
+    , CautiousT(..)
+    , CautiousS
+    , cautiousWarning
+    , cautiousWarningM
+    , cautiousErrorIfNothing
+    , cautiousWarningIfNothing
+    , prettyPrintWarn
+    , cautiousError
+    , prettyPrintErr
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: Again, no need to export Cautious.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 6]
```diff
@@ -89,3 +105,48 @@ instance ToJSON Repo where
...
+
+data Err =
+    NoInitFile
+    deriving (Show, Eq, Generic)
+
+instance ToJSON Err
+
+instance FromJSON Err
+
+prettyPrintErr :: Err -> String
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: It's not pretty printing if you're not printing.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 7]
```diff
@@ -89,3 +105,48 @@ instance ToJSON Repo where
...
+prettyPrintErr NoInitFile =
+    "Error: the wolf repository is not or poorly initialised."
+
+prettyPrintProblem :: Problem -> String
+prettyPrintProblem (WarnMissingNoteIndex pu) =
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: Why is this a warning? If a note doesn't have any notes, it doesn't need a note index.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 8]
```diff
@@ -89,3 +105,48 @@ instance ToJSON Repo where
...
+    "Warning: " ++ show pu ++ " has no NoteIndex."
+prettyPrintProblem (WarnMissingNote nu) =
+    "Warning: " ++ show nu ++ " has no note."
+prettyPrintProblem (WarnMissingRelevantPerson pu nu) =
+    "Warning: The note " ++
+    show nu ++ " should mention the person " ++ show pu ++ "."
+prettyPrintProblem (WarnMissingRelevantNote nu pu) =
+    "Warning: The person " ++
+    show pu ++ " should mention the note " ++ show nu ++ " in his noteIndex."
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: Instead of writing `X should Y`, which sounds like an opinion.
Write `X does not Y` (while that's what we expected).

File: wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs [hunk 9]
```diff
@@ -61,3 +61,20 @@ instance GenValid Repo where
             fmap M.fromList $ forM noteUuids $ \uuid -> (,) uuid <$> genValid
         repoSuggestions <- genValid
         pure Repo {..}
+
+instance (GenUnchecked a, GenUnchecked w, GenUnchecked e) =>
```
Comment (NorfairKing) [inline wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs]: This should really be in the library, but since you're pinning its version, that's fine.

File: wolf-data/src/Wolf/Data/Cautious.hs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+       (MonadIO m, MonadReader DataSettings m)
+    => NoteUuid
+    -> CautiousS m (Maybe Note)
+readNoteCautious nu = do
+    maybeNote <- lift $ readNote nu
+    case maybeNote of
+        Nothing -> cautiousWarning [WarnMissingNote nu] Nothing
+        Just note -> do
+            foldl'
+                (\warns pu -> warns >> checkPersonNoteRelation nu pu)
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Cautious.hs]: Have a look at `foldM` to see if that works better than this `foldl'`.

**Review by NorfairKing — CHANGES_REQUESTED** [review]
Preliminary review. Didn't continue because there was too much that needs to be changed already.

File: wolf-data/src/Wolf/Data/Export/Types.hs (hunk 6 above)
Comment (CSVdB) [reply wolf-data/src/Wolf/Data/Export/Types.hs]: I already changed it. However:
https://hackage.haskell.org/package/haskell-src-exts-1.20.2/docs/Language-Haskell-Exts-Pretty.html#v:prettyPrint
That is your interpretation of the word "print".

File: wolf-data/src/Wolf/Data/Export/Types.hs (hunk 7 above)
Comment (CSVdB) [reply wolf-data/src/Wolf/Data/Export/Types.hs]: If you would look at where it's used, you'd know it's not a warning I throw because just if a noteindex doesn't. Instead, I look up notes, look at the relevant people corresponding to these notes. Those people should have a noteindex, and the noteindex should contain the aforementioned note.
I'll add the NoteUuid to the warning, if that makes it clearer.

File: wolf-data/src/Wolf/Data/Export/Types.hs (hunk 8 above)
Comment (CSVdB) [reply wolf-data/src/Wolf/Data/Export/Types.hs]: I changed it to your pleasing.
However, I would like to know why I need to do this.
"Should" means (in my opinion) "it would be better if it did, but it's not absolutely necessary". This, imo, expresses exactly what a "warning" means.

File: wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs (hunk 9 above)
Comment (CSVdB) [reply wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs]: Do you want me to add a genvalidity dependency to a library?
It is in the cautious repo, just in the tests...

File: wolf-data/src/Wolf/Data/Cautious.hs (hunk 3 above)
Comment (CSVdB) [reply wolf-data/src/Wolf/Data/Cautious.hs]: First of all: I already made the change.

I'm not sure why I'd write a function which can do less (only take Problem, not [Problem], which can do strictly more), especially since it doesn't even clean up the code. Every place in the codebase where cautiousWarning is used, I now need [] in the first argument, and cautiousProblem would need () around the first argument. Can you explain me why you think this is better?

Also, this function gives a single warning, just not necessarily a single Problem.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 11]
```diff
@@ -75,7 +67,60 @@ instance Validity Repo where
...
             , repoSuggestions <?!> "repoSuggestions"
+            , mconcat $
+              itsNotesMentionPerson repoNoteIndices repoNotes <$>
+              (snd <$> indexTuples repoPersonIndex) -- If pu refers to nu, nu refers to pu
+            , mconcat $
+              itsPeopleMentionNote repoNoteIndices repoNotes <$>
+              S.toList (noteIndexSet repoNoteIndex)
+            -- If nu refers to pu, pu refers to nu
             ]
+    isValid = null . unValidation . validate
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: that's `isValidByValidating`.
But this is removed in recent versions of validity, so we better upgrade at some point.
Not in this PR though.

File: wolf-data/src/Wolf/Data/Prompt.hs [hunk 12]
```diff
@@ -1,4 +1,4 @@
-module Wolf.Cli.Prompt
+module Wolf.Data.Prompt
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Prompt.hs]: What?! Why?!

No No No!

Promptiing makes no sense for other reverse dependencies. Leave this code where it belongs.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 13]
```diff
@@ -89,3 +143,65 @@ instance ToJSON Repo where
             , "notes" .= repoNotes
             , "suggestions" .= repoSuggestions
             ]
+
+type Warn = [Problem]
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: These are rather badly named.
I guess this should be `ExportWarning` and `ExportProblem` I can't tell, but I guess so.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 14]
```diff
@@ -89,3 +143,65 @@ instance ToJSON Repo where
...
+    = WarnMissingNoteIndex PersonUuid
+                           NoteUuid
+    | WarnMissingNote NoteUuid
+    | WarnMissingRelevantPerson PersonUuid
+                                NoteUuid
+    | WarnMissingRelevantNote NoteUuid
+                              PersonUuid
+    deriving (Show, Eq, Generic)
+
+instance ToJSON Problem
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: I'm not sure you need these.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 15]
```diff
@@ -89,3 +143,65 @@ instance ToJSON Repo where
...
+
+instance FromJSON Problem
+
+data Err =
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: name

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 16]
```diff
@@ -89,3 +143,65 @@ instance ToJSON Repo where
...
+    NoInitFile
+    deriving (Show, Eq, Generic)
+
+instance ToJSON Err
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: unnecessary?

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 17]
```diff
@@ -89,3 +143,65 @@ instance ToJSON Repo where
...
+
+instance FromJSON Err
+
+type CautiousS = CautiousT Warn Err
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: What does the `S` stand for?. Another bad name.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 18]
```diff
@@ -89,3 +143,65 @@ instance ToJSON Repo where
...
+
+cautiousProblem :: Monad m => Problem -> a -> CautiousS m a
+cautiousProblem p = cautiousWarning [p]
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: nit: use pointfree notation.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 19]
```diff
@@ -89,3 +143,65 @@ instance ToJSON Repo where
...
+
+cautiousProblemIfNothing ::
+       Monad m => Problem -> Maybe a -> CautiousS m (Maybe a)
+cautiousProblemIfNothing p = cautiousWarningIfNothing [p]
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: It seems all of these should be in the cautious library instead of here.

File: wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs [hunk 20]
```diff
@@ -19,6 +21,8 @@ import Wolf.Data.NoteIndex.Types.Gen
 import Wolf.Data.People.Types.Gen ()
 import Wolf.Data.Suggestion.Types.Gen ()
 
+import Cautious.Cautious
```
Comment (NorfairKing) [inline wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs]: Nit: these go _before_ the `Wolf` imports.

File: wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs [hunk 21]
```diff
@@ -29,22 +33,15 @@ instance GenValid Repo where
...
-                    b <- frequency [(3, pure False), (1, pure True)]
-                    if b
-                        then pure ix
-                        else do
-                            a <- genValid
-                            go (addIndexEntry a puuid ix) puuid
+            let gen ix puuid = do
+                    a <- genValid
+                    go (addIndexEntry a puuid ix) puuid
+                go ix puuid = frequency [(3, gen ix puuid), (1, pure ix)]
```
Comment (NorfairKing) [inline wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs]: Oh, smart, but why did you change this? Seems like a refactor that doesn't belong here.

File: wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs [hunk 22]
```diff
@@ -29,22 +33,15 @@ instance GenValid Repo where
...
             in foldM go newIndex eps
         -- For some people, make a person entry.
         repoPersonEntries <-
             fmap (M.fromList . catMaybes) $
-            forM eps $ \p -> do
-                b <- genValid
-                if b
-                    then (Just . (,) p) <$> genValid
-                    else pure Nothing
+            forM eps $ \p -> oneof [pure Nothing, (Just . (,) p) <$> genValid]
```
Comment (NorfairKing) [inline wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs]: same

File: wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs [hunk 23]
```diff
@@ -57,7 +54,47 @@ instance GenValid Repo where
...
+        let repoNotes =
+                foldl'
+                    updateNotes
+                    repoNotesWithWrongPeople
+                    (M.toList repoNoteIndices)
         repoSuggestions <- genValid
         pure Repo {..}
+
+rmRelevantPeople :: Note -> Note
+rmRelevantPeople note = note {noteRelevantPeople = S.empty}
```
Comment (NorfairKing) [inline wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs]: rm -> remove?

File: wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs [hunk 24]
```diff
@@ -57,7 +54,47 @@ instance GenValid Repo where
...
+    M.adjust
+        (\note ->
+             note {noteRelevantPeople = S.insert pu $ noteRelevantPeople note})
+        nu
+        noteMap
+
+instance (GenUnchecked a, GenUnchecked w, GenUnchecked e) =>
+         GenUnchecked (Cautious w e a)
+
+instance (GenValid a, GenValid w, GenValid e) => GenValid (Cautious w e a)
```
Comment (NorfairKing) [inline wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs]: This should be in `cautious-gen`, really.

File: wolf-test-utils/src/Wolf/Data/TestUtils.hs [hunk 25]
```diff
@@ -14,6 +14,9 @@ import Wolf.Data
 
 import Wolf.Data.Gen ()
 
+import Cautious.Cautious
+import Cautious.CautiousT
```
Comment (NorfairKing) [inline wolf-test-utils/src/Wolf/Data/TestUtils.hs]: Shouldn't `CautiousT` export `Cautious`?

File: wolf-test-utils/test/Wolf/Data/Export/TypesSpec.hs [hunk 26]
```diff
@@ -8,10 +8,16 @@ import TestImport
 
 import Wolf.Data.Export.Types
 
+import Wolf.Data.Export.Types.Gen ()
```
Comment (NorfairKing) [inline wolf-test-utils/test/Wolf/Data/Export/TypesSpec.hs]: This should be exported from `Wolf.Data.Gen`, not added here.

**Review by NorfairKing — CHANGES_REQUESTED** [review]

File: wolf-data/src/Wolf/Data/Prompt.hs (hunk 12 above)
Comment (CSVdB) [reply wolf-data/src/Wolf/Data/Prompt.hs]: The reason I did this, is to avoid wolf-data to depend on wolf-cli. The options are:
1. Put the code in wolf data
2. Make wolf data dependent on wolf-cli (which means there's a cyclic dependency, probably causing problems in the long run, as you saw with validity and zifter)
3. Duplicate the code, and maintain it in two places
4. Put the code in a separate repository, and put that on hackage and stackage.

In my opinion, option 2 is the best one. What do you think?

File: wolf-data/src/Wolf/Data/Export/Types.hs (hunk 19 above)
Comment (CSVdB) [reply wolf-data/src/Wolf/Data/Export/Types.hs]: The cautious library has no "problem".

File: wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs (hunk 23 above)
Comment (CSVdB) [reply wolf-test-utils/src/Wolf/Data/Export/Types/Gen.hs]: The reason I used this abbreviation is because everyone in tech is familiar with it from the command line command. I changed it though.

File: wolf-data/src/Wolf/Data/Prompt.hs (hunk 12 above)
Comment (NorfairKing) [reply wolf-data/src/Wolf/Data/Prompt.hs]: Why do you need prompt in data? If you do, something is wrong.

File: wolf-cli/src/Wolf/Cli/Command/Export.hs [hunk 27]
```diff
@@ -14,9 +15,19 @@ import Wolf.Data
...
     withInitCheck_ $ do
-        e <- exportRepo
-        liftIO . LB8.putStrLn $ encodePretty e
+        e <- runCautiousT exportRepo
+        case e of
+            CautiousWarning [] repo -> liftIO . LB8.putStrLn $ encodePretty repo
+            CautiousWarning w repo ->
+                liftIO . LB8.putStrLn $
+                LB8.unlines
+                    [encodePretty repo, "The warnings are:", encodePretty w]
```
Comment (NorfairKing) [inline wolf-cli/src/Wolf/Cli/Command/Export.hs]: This is weird. Either show the warnings for humans or put everything in one JSON object.
I prefer the former.

File: wolf-data/src/Wolf/Data/Cautious.hs [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+                         (ExportWarningMissingRelevantPerson pu nu)
+                         ()
+
+getPersonNoteIndexCautious ::
+       (MonadIO m, MonadReader DataSettings m)
+    => PersonUuid
+    -> CautiousExport m (Maybe NoteIndex)
+getPersonNoteIndexCautious pu = do
+    noteIndex <- lift $ getPersonNoteIndex pu
+    case noteIndex of
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Cautious.hs]: `forM noteIndex`

File: wolf-data/src/Wolf/Data/Export.hs [hunk 29]
```diff
@@ -5,51 +5,61 @@ module Wolf.Data.Export
...
+    initData <- cautiousErrorIfNothing mid NoInitFile
+    mi <- lift getIndexWithDefault
+    people <- lift getPersonUuids -- These are collected from the people directory
+    entries <- lift $ mKeyed getPersonEntry people
+    noteIndex <- lift getNoteIndex
+    mNoteIxs <- getMapCautious getPersonNoteIndexCautious people
+    noteUuids <- lift getNoteUuids
+    notes <- getMapCautious readNoteCautious noteUuids
+    sugs <- lift readAllSuggestions
+    pure
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export.hs]: Don't we also want to show a warning if the resulting `Repo` is not `isValid`?
Maybe make an extra warning called `ExportWarningRepoInvalid Repo` so that you can show its pretty validation.

File: wolf-data/src/Wolf/Data/Export/Types.hs [hunk 30]
```diff
@@ -89,3 +143,69 @@ instance ToJSON Repo where
...
+        , " mentions him/her."
+        ]
+prettyShowExportProblem (ExportWarningMissingNote nu) =
+    "ExportWarninging: " ++ show nu ++ " has no note."
+prettyShowExportProblem (ExportWarningMissingRelevantPerson pu nu) =
+    "ExportWarninging: The note " ++
+    show nu ++ " does not mention the person " ++ show pu ++ "."
+prettyShowExportProblem (ExportWarningMissingRelevantNote nu pu) =
+    "ExportWarninging: The person " ++
+    show pu ++ " does not mention the note " ++ show nu ++ " in his noteIndex."
```
Comment (NorfairKing) [inline wolf-data/src/Wolf/Data/Export/Types.hs]: Person is not male, person is neutral.
-> 'Their noteIndex'

**Review by NorfairKing — APPROVED** [review]
I'm happy with this as-is. Up to you whether you want to fix the things I commented on.

File: wolf-data/src/Wolf/Data/Export.hs (hunk 29 above)
Comment (CSVdB) [reply wolf-data/src/Wolf/Data/Export.hs]: I thought about this carefully, and I don't believe this should be a warning.
The entire point (in my opinion) of the "export" command is to export a repository which is as big as possible while only representing valid things which are in your filesystem. If the resulting repo is invalid, this task has failed completely and hence, in my opinion, we should do nothing more or less than explain the user as well as possible what is wrong about his repo.

Given this, I wrote a wrapper InvalidRepoMessage around Strings, and made a smart constructor which assures that you can only create an InvalidRepoMessage (used to create an ExportErrorRepoInvalid) through a function Repo -> Either InvalidRepoMessage Repo.

**Comment by CSVdB** [comment]
All comments are fixed except for the one about the invalid repo warning, as explained in that conversation.
